### PR TITLE
Use a new jinja2 templating process to update library to latest unicode (14.0)

### PR DESCRIPTION
**Pull request by jquast**
- update `tox.ini`, `requirements*` files
  - use `pip-compile` with "requirements.in" files to generate "requirements.txt" files
  - use older testing tools for older python versions (requirements-tests38.in and requirements-tests39.in)
  - use python3.10 for most build steps
- bin/update-tables.py changes,
  - use https
  - use jinja2 (`.j2`) templating to generate code and documentation,
    - easier to support code generation of other languages
  - unreliable connection support using tenacity library to retry on connection timeout, etc.
  - optional `--check-last-modified` will compare last-modified  for redownload
    - a unicode.org data file was only once updated for a botched release
- `for wcwidth/table*.py` files
  - uses new template system to generate new table data files for unicode version 14
  - when a description of unicode is single codepoint, describe only as 'A' instead of 'A..A'

### CODE REVIEW

File: bin/update-tables.py [hunk 1]
```diff
@@ -1,332 +1,446 @@
...
 
+JINJA_ENV = jinja2.Environment(
+    loader=jinja2.FileSystemLoader(os.path.join(PATH_UP, 'code_templates')),
+    keep_trailing_newline=True)
+UTC_NOW = datetime.datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S UTC")
 
-TableDef = collections.namedtuple('table', ['version', 'date', 'values'])
+CONNECT_TIMEOUT = int(os.environ.get('CONNECT_TIMEOUT', '10'))
+FETCH_BLOCKSIZE = int(os.environ.get('FETCH_BLOCKSIZE', '4096'))
+MAX_RETRIES = int(os.environ.get('MAX_RETRIES', '10'))
```
Comment (GalaxySnail) [inline bin/update-tables.py:59]: 10 retries may be too many here, the tenth retry will wait for `0.1 * (2 ** (10 - 1)) == 51.2` seconds. 6 or 7 retries are enough.

File: tox.ini [hunk 2]
```diff
@@ -60,20 +58,47 @@ precision = 1
...
+deps = -r requirements-tests37.txt
+
+[testenv:py37]
+deps = -r requirements-tests37.txt
+
+[testenv:py36]
+deps = -r requirements-tests36.txt
 
 [testenv:update]
+basepython = python3.10
```
Comment (GalaxySnail) [inline tox.ini:95]: Shall we update to python3.11 here?

**Review by jquast — COMMENTED** [review]
thanks for the help, I'll merge & release in ~8 hours or so, I just want to be sure I am home to address any issues in release, and maybe address a few small open issues as well, #22, #24, #55, #56.

With any luck I will be able to work on Cythonification tomorrow, I found this stack overflow solution that addresses my concern, that we can include a cython or c extension as *optional* and fallback to pure python if it is not able to compile, https://stackoverflow.com/questions/41778153/compiling-an-optional-cython-extension-only-when-possible-in-setup-py

**Comment by GalaxySnail** [comment]
> thanks for the help, I'll merge & release in ~8 hours or so, I just want to be sure I am home to address any issues in release, and maybe address a few small open issues as well, #22, #24, #55, #56.

Thanks for your efforts! Everything LGTM. Updating to python 3.11 in tox.ini can be done in another PR. 

> With any luck I will be able to work on Cythonification tomorrow, I found this stack overflow solution that addresses my concern, that we can include a cython or c extension as _optional_ and fallback to pure python if it is not able to compile, https://stackoverflow.com/questions/41778153/compiling-an-optional-cython-extension-only-when-possible-in-setup-py

This anwser looks really old, we have `pyproject.toml` and [PEP 517](https://peps.python.org/pep-0517/) nowadays. The [`get_requires_for_build_wheel` hook](https://peps.python.org/pep-0517/#get-requires-for-build-wheel) and an [in-tree build backend](https://peps.python.org/pep-0517/#in-tree-build-backends) will be helpful. Please see: https://github.com/psycopg/psycopg/issues/471#issuecomment-1368531676. Also, I think it can be done in another PR.

**Review by GalaxySnail — APPROVED** [review]

**Pull request merged**